### PR TITLE
Support Full Join

### LINKED ISSUE #287: Add support for FULL JOIN
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Add support for FULL JOIN.

**Describe the solution you'd like**
Implementation shoud be pretty straightforward: it is the combination of the already implemented left and right join.
We should also update some documentation.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by Dandandan**
# Which issue does this PR close?

Closes #287

 # Rationale for this change

Full join completes the support for the basic join types in DataFusion.

# What changes are included in this PR?

Add support for full join, via support for left/right join.

# Are there any user-facing changes?

No

### CODE REVIEW

File: ballista/rust/core/src/serde/physical_plan/to_proto.rs [hunk 1]
```diff
@@ -132,6 +132,7 @@ impl TryInto<protobuf::PhysicalPlanNode> for Arc<dyn ExecutionPlan> {
                 JoinType::Inner => protobuf::JoinType::Inner,
                 JoinType::Left => protobuf::JoinType::Left,
                 JoinType::Right => protobuf::JoinType::Right,
+                JoinType::Full => protobuf::JoinType::Right,
```
Comment (alamb) [inline ballista/rust/core/src/serde/physical_plan/to_proto.rs:135]: This looks like it is a typo perhaps

```suggestion
                JoinType::Full => protobuf::JoinType::Full,
```

File: datafusion/src/physical_plan/hash_join.rs [hunk 2]
```diff
@@ -1410,6 +1413,45 @@ mod tests {
...
+
+        let join = join(left, right, on, &JoinType::Full).unwrap();
+
+        let columns = columns(&join.schema());
+        assert_eq!(columns, vec!["a1", "b1", "c1", "a2", "c2"]);
+
+        let stream = join.execute(0).await.unwrap();
+        let batches = common::collect(stream).await.unwrap();
+
+        let expected = vec![
```
Comment (alamb) [inline datafusion/src/physical_plan/hash_join.rs:1438]: This doesn't look like the correct answer to me. FULL OUTER JOIN basically keeps tuples from boths sides that don't match the join criteria.

So we should have 2 rows for `b1=4' and `b1=5` and then 1 row for tuple on the left that has no match with the right, `b1=7` and 1 row for the tuple on the right that has no match with the left `b1=6`

And indeed  postgres returns 4 rows as well


```
alamb=# select * from tleft;
 a1 | b1 | c1 
----+----+----
  1 |  4 |  7
  2 |  5 |  8
  3 |  7 |  9
(3 rows)

alamb=# select * from tright;
 a2 | b2 | c2 
----+----+----
 10 |  4 | 70
  2 |  5 | 80
  3 |  6 | 90
(3 rows)

alamb=# select * from tleft FULL JOIN tright ON (tleft.b1 = tright.b2);
 a1 | b1 | c1 | a2 | b2 | c2 
----+----+----+----+----+----
  1 |  4 |  7 | 10 |  4 | 70
  2 |  5 |  8 |  2 |  5 | 80
    |    |    |  3 |  6 | 90
  3 |  7 |  9 |    |    |   
(4 rows)
```

File: datafusion/src/physical_plan/hash_join.rs [hunk 3]
```diff
@@ -1515,6 +1589,43 @@ mod tests {
...
+        assert_eq!(columns, vec!["a1", "b1", "c1", "a2", "c2"]);
+
+        let stream = join.execute(0).await?;
+        let batches = common::collect(stream).await?;
+
+        let expected = vec![
+            "+----+----+----+----+----+",
+            "| a1 | b1 | c1 | a2 | c2 |",
+            "+----+----+----+----+----+",
+            "|    |    |    | 30 | 90 |",
```
Comment (alamb) [inline datafusion/src/physical_plan/hash_join.rs:1618]: this row should also have `b1=6` I think

**Review by alamb — COMMENTED** [review]
Thanks @Dandandan  -- this is great. Something doesn't quite look right with the answers to me. 

It might also be worth a quick test to verify that the `FULL OUTER JOIN` syntax works the same as `FULL JOIN`

File: datafusion/src/physical_plan/hash_join.rs (hunk 2 above)
Comment (Dandandan) [reply datafusion/src/physical_plan/hash_join.rs:1438]: The test here tests having two (equivalent) batches at the right side.
So they will produce two rows for the right rows and one for the left row that doesnt match the right row.

File: datafusion/src/physical_plan/hash_join.rs (hunk 3 above)
Comment (Dandandan) [reply datafusion/src/physical_plan/hash_join.rs:1618]: The `b1` column is here from the left side. I can make the test a bit easier to follow.

I believe this should be covered by the work that @houqp is working on to support qualified names in tables (and disambiguating using the names).

Comment (Dandandan) [reply datafusion/src/physical_plan/hash_join.rs:1618]: I changed the test to use `b2` as column so you can see better what's going on:

```
            "+----+----+----+----+----+----+",
            "| a1 | b1 | c1 | a2 | b2 | c2 |",
            "+----+----+----+----+----+----+",
            "|    |    |    | 30 | 6  | 90 |",
            "| 1  | 4  | 7  | 10 | 4  | 70 |",
            "| 2  | 5  | 8  | 20 | 5  | 80 |",
            "| 3  | 7  | 9  |    |    |    |",
            "+----+----+----+----+----+----+",
```

File: datafusion/src/physical_plan/hash_join.rs (hunk 2 above)
Comment (Dandandan) [reply datafusion/src/physical_plan/hash_join.rs:1438]: To me the results look correct given two equivalent right batches. I included `b2` as column here as well to make it a bit more obvious:

```
            "+----+----+----+----+----+----+",
            "| a1 | b1 | c1 | a2 | b2 | c2 |",
            "+----+----+----+----+----+----+",
            "|    |    |    | 30 | 6  | 90 |",
            "|    |    |    | 30 | 6  | 90 |",
            "| 1  | 4  | 7  | 10 | 4  | 70 |",
            "| 1  | 4  | 7  | 10 | 4  | 70 |",
            "| 2  | 5  | 8  | 20 | 5  | 80 |",
            "| 2  | 5  | 8  | 20 | 5  | 80 |",
            "| 3  | 7  | 9  |    |    |    |",
            "+----+----+----+----+----+----+",
```

**Comment by Dandandan** [comment]
@alamb I updated the tests and addressed your comments

**Review by jorgecarleitao — APPROVED** [review]
Went through the PR and it looks great to me. The implementation is what I would expect, and the tests cover well all cases. Really great work, @Dandandan 

Two other notes:

* the amount of changes to enable this was really small, which IMO shows that we already have a good design around joins. Thank you all for making this happen, it is great when we can add such a feature with so little code 💯 
* Ballista proto also updated! 💯

**Review by houqp — APPROVED** [review]
Good work :+1:

File: datafusion/tests/sql.rs [hunk 4]
```diff
@@ -1342,6 +1342,11 @@ async fn outer_join() -> Result<()> {
         vec!["44", "d", "x"],
     ];
     assert_eq!(expected, actual);
+
+    let sql = "SELECT t1_id, t1_name, t2_name FROM t1 FULL OUTER JOIN t2 ON t1_id = t2_id ORDER BY t1_id";
```
Comment (alamb) [inline datafusion/tests/sql.rs:1346]: 👍

File: datafusion/src/physical_plan/hash_join.rs [hunk 5]
```diff
@@ -1590,36 +1591,36 @@ mod tests {
...
         let expected = vec![
-            "+----+----+----+----+----+",
-            "| a1 | b1 | c1 | a2 | c2 |",
-            "+----+----+----+----+----+",
-            "|    |    |    | 30 | 90 |",
-            "| 1  | 4  | 7  | 10 | 70 |",
-            "| 2  | 5  | 8  | 20 | 80 |",
-            "| 3  | 7  | 9  |    |    |",
-            "+----+----+----+----+----+",
+            "+----+----+----+----+----+----+",
```
Comment (alamb) [inline datafusion/src/physical_plan/hash_join.rs:1616]: I see -- the output was only showing `b1` from `left` -- that was very confusing. Thank you for making it clearer

**Review by alamb — APPROVED** [review]
I agree -- this looks great. Thank you so much @Dandandan

**Comment by Dandandan** [comment]
Thanks all for the reviews! :+1::+1::+1:

@jorgecarleitao yeah really happy that this only involved adding some pattern matches and we could just reuse the code for right / left join!

**Pull request merged**